pallets/foreign-investments/src/impls/mod.rs

```diff
@@ -0,0 +1,1042 @@
+	///
+	/// Returns a tuple of the total swap order amount as well as potentially
+	/// altered invest and redeem states. Any returning tuple element which is
+	/// `None`, reflects that no change is required for this element. Else, it
+	/// needs to be applied to the storage.
+	///
+	/// NOTE: Required since there exists at most one swap per `(AccountId,
+	/// InvestmentId)` pair whereas investments and redemptions can both mutate
+	/// orders. Assume, as a result of an `InvestState` transition, a token swap
+	/// order into pool currency is initialized. Then, as a result of a
+	/// `RedeemState` transition, a token swap order into return currency is
+	/// needed. This handler resolves the _merge conflict_ in situations where
+	/// the reason to create/update a swap order does not match the previous
+	/// reason.
+	///
+	/// * Is noop, if the the current reason equals the previous one.
+	/// * If both states are swapping into return currency, i.e. their invest
+	///   and redeem states include `ActiveSwapIntoReturnCurrency`, the states
+	///   stay the same. However the total order amount needs to be updated by
+	///   summing up both swap order amounts.
+	/// * If the `InvestState` includes swapping into pool currency, i.e.
+	///   `ActiveSwapIntoPoolCurrency`, whereas the `RedeemState` is swapping
+	///   into the opposite direction, i.e. `ActiveSwapIntoReturnCurrency`, we
+	///   need to resolve the delta between both swap order amounts and update
+	///   the states accordingly.
+	#[allow(clippy::type_complexity)]
+	fn handle_concurrent_swap_orders(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		swap_order_id: T::TokenSwapOrderId,
+		reason: TokenSwapReason,
+	) -> Result<
+		(
+			Option<Swap<T::Balance, T::CurrencyId>>,
+			Option<InvestState<T::Balance, T::CurrencyId>>,
+			Option<RedeemState<T::Balance, T::CurrencyId>>,
+		),
+		DispatchError,
+	> {
+		let last_reason = ForeignInvestmentInfo::<T>::get(swap_order_id)
+			.ok_or(Error::<T>::InvestmentInfoNotFound)?
+			.last_swap_reason
+			.ok_or(Error::<T>::TokenSwapReasonNotFound)?;
+
+		// Exit early if both reasons match, i.e. we would not override any opposite
+		// swap order
+		if last_reason == reason {
+			return Ok((None, None, None));
+		}
+
+		// Read states from storage and determine amounts
+		let invest_state = InvestmentState::<T>::get(who, investment_id).unwrap_or_default();
+		let redeem_state = RedemptionState::<T>::get(who, investment_id).unwrap_or_default();
+		let invest_swap_amount = invest_state
+			.get_active_swap()
+			.map(|s| s.amount)
+			.unwrap_or_default();
+		let redeem_swap_amount = redeem_state
+			.get_active_swap()
+			.map(|s| s.amount)
+			.unwrap_or_default();
+		let resolved_amount = invest_swap_amount.min(redeem_swap_amount);
+		// safe because max >= min, equals zero if both amounts equal
+		let swap_amount_opposite_direction =
+			invest_swap_amount.max(redeem_swap_amount) - resolved_amount;
```

Comment:
I found this pattern before too. I think would be cool to have a utility to make an absolute diff, maybe in `cfg_traits::ops`?.